hwy/contrib/unroller/unroller_test.cc

```diff
@@ -0,0 +1,295 @@
+// clang-format on
+
+#define UNUSED(x) (void)(x)
+
+HWY_BEFORE_NAMESPACE();
+namespace hwy {
+namespace HWY_NAMESPACE {
+
+template <typename T>
+T SimpleDot(const T* pa, const T* pb, size_t num) {
+  T sum = 0;
+  for (size_t i = 0; i < num; ++i) {
+    sum += pa[i] * pb[i];
+  }
+  return sum;
+}
+
+template <typename T>
+T SimpleAcc(const T* pa, size_t num) {
+  T sum = 0;
+  for (size_t i = 0; i < num; ++i) {
+    sum += pa[i];
+  }
+  return sum;
+}
+
+template <typename T>
+T SimpleMin(const T* pa, size_t num) {
+  T min = std::numeric_limits<T>::max();
+  for (size_t i = 0; i < num; ++i) {
+    if(min > pa[i])
+        min = pa[i];
+  }
+  return min;
+}
+
+template <typename T>
+struct accv : UnrollerUnit<accv<T>, T, T> {
+  using TT = hn::ScalableTag<T>;
+  inline hn::Vec<TT> Func(ptrdiff_t const& HWY_RESTRICT idx,
+                          hn::Vec<TT> const& HWY_RESTRICT x,
+                          hn::Vec<TT> const& HWY_RESTRICT y) const {
+    return x + y;
+  }
+
+  inline ptrdiff_t StoreImpl(ptrdiff_t const& HWY_RESTRICT idx,
+                             T* HWY_RESTRICT to, const hn::Vec<TT> x) const {
+    // no stores in a reducer
+    return 0;
+  }
+
+  inline ptrdiff_t MaskStoreImpl(const ptrdiff_t idx, T* HWY_RESTRICT to,
+                                 hn::Vec<TT> const& HWY_RESTRICT x,
+                                 ptrdiff_t const& HWY_RESTRICT places) const {
+    // no stores in a reducer
+    return 0;
+  }
+
+  inline ptrdiff_t ReduceImpl(hn::Vec<TT> const& HWY_RESTRICT x,
+                              T* HWY_RESTRICT to) const {
+    const hn::ScalableTag<T> d;
+    (*to) = hn::ReduceSum(d, x);
+    return 1;
+  }
+
+  inline void ReduceImpl(hn::Vec<TT> const& HWY_RESTRICT x0,
+                         hn::Vec<TT> const& HWY_RESTRICT x1,
+                         hn::Vec<TT> const& HWY_RESTRICT x2,
+                         hn::Vec<TT>& HWY_RESTRICT y) const {
+    y = y + x0 + x1 + x2;
+  }
+};
+
+template <typename T>
+struct minv : UnrollerUnit<minv<T>, T, T> {
+  using TT = hn::ScalableTag<T>;
+  TT d;
+
+  inline hn::Vec<TT> Func(ptrdiff_t const& HWY_RESTRICT idx,
+                          hn::Vec<TT> const& HWY_RESTRICT x,
+                          hn::Vec<TT> const& HWY_RESTRICT y) const {
+    return hn::Min(y, x);
+  }
+
+  inline hn::Vec<TT> YInitImpl() const {
+    return hn::Set(d, std::numeric_limits<T>::max());
+  }
+
+  inline hn::Vec<TT> MaskLoadImpl(ptrdiff_t const& HWY_RESTRICT idx,
+                                  T* HWY_RESTRICT from,
+                                  ptrdiff_t const& HWY_RESTRICT places) const {
+    const hn::ScalableTag<ptrdiff_t> di;
+    using TI = hn::TFromD<decltype(di)>;
+    auto mask = hn::FirstN(d, static_cast<size_t>(places));
+    auto maskneg = hn::Not(hn::FirstN(
+        d, static_cast<size_t>(places +
+                               static_cast<ptrdiff_t>(
+                                   UnrollerUnit<minv<T>, T, T>::UnitLanes()))));
```

Comment:
Should we encourage a typedef called Base in these classes? Or perhaps a constexpr size_t kMaxLanes at the beginning of the class next to the inheritance?

Reply:
I'm thinking about the right way to handle this.